primitives/storage/src/lib.rs

```diff
@@ -126,180 +129,201 @@ pub mod well_known_keys {
-impl OwnedChildInfo {
-	/// Instantiates info for a default child trie.
-	pub fn new_default(unique_id: Vec<u8>) -> Self {
-		OwnedChildInfo::Default(OwnedChildTrie {
-			data: unique_id,
+impl ChildType {
+	/// Try to get a child type from its `u32` representation.
+	pub fn new(repr: u32) -> Option<ChildType> {
+		Some(match repr {
+			r if r == ChildType::ParentKeyId as u32 => ChildType::ParentKeyId,
+			_ => return None,
 		})
 	}
 
-	/// Try to update with another instance, return false if both instance
-	/// are not compatible.
-	pub fn try_update(&mut self, other: ChildInfo) -> bool {
-		match self {
-			OwnedChildInfo::Default(owned_child_trie) => owned_child_trie.try_update(other),
+	/// Produce a prefixed key for a given child type.
+	fn new_prefixed_key(&self, key: &[u8]) -> Vec<u8> {
+		let parent_prefix = self.parent_prefix();
+		let mut result = Vec::with_capacity(parent_prefix.len() + key.len());
+		result.extend_from_slice(parent_prefix);
+		result.extend_from_slice(key);
+		result
+	}
+
+	/// Prefixes a vec with the prefix for this child type.
+	fn do_prefix_key(&self, key: &mut Vec<u8>) {
+		let parent_prefix = self.parent_prefix();
+		let key_len = key.len();
+		if parent_prefix.len() > 0 {
+			key.resize(key_len + parent_prefix.len(), 0);
+			key.copy_within(..key_len, parent_prefix.len());
+			key[..parent_prefix.len()].copy_from_slice(parent_prefix);
 		}
 	}
 
-	/// Get `ChildInfo` reference to this owned child info.
-	pub fn as_ref(&self) -> ChildInfo {
+	/// Returns the location reserved for this child trie in their parent trie if there
+	/// is one.
+	fn parent_prefix(&self) -> &'static [u8] {
 		match self {
-			OwnedChildInfo::Default(OwnedChildTrie { data })
-				=> ChildInfo::Default(ChildTrie {
-					data: data.as_slice(),
-				}),
+			&ChildType::ParentKeyId => DEFAULT_CHILD_TYPE_PARENT_PREFIX,
 		}
 	}
 }
 
 /// A child trie of default type.
-/// Default is the same implementation as the top trie.
-/// It share its trie node storage with any kind of key,
-/// and its unique id needs to be collision free (eg strong
-/// crypto hash).
-#[derive(Clone, Copy)]
-pub struct ChildTrie<'a> {
-	/// Data containing unique id.
-	/// Unique id must but unique and free of any possible key collision
-	/// (depending on its storage behavior).
-	data: &'a[u8],
-}
-
-/// Owned version of default child trie `ChildTrie`.
+/// It uses the same default implementation as the top trie,
+/// top trie being a child trie with no keyspace and no storage key.
+/// Its keyspace is the variable (unprefixed) part of its storage key.
+/// It shares its trie nodes backend storage with every other
+/// child trie, so its storage key needs to be a unique id
+/// that will be use only once.
 #[derive(Debug, Clone)]
 #[cfg_attr(feature = "std", derive(PartialEq, Eq, Hash, PartialOrd, Ord))]
-pub struct OwnedChildTrie {
-	/// See `ChildTrie` reference field documentation.
+pub struct ChildTrieParentKeyId {
+	/// Data is the full prefixed storage key.
```

Comment:
If I read correctly throughout this module data is considered without prefix.
```suggestion
	/// Data is the storage key without prefix.
```